### DISCUSSION TITLE
CNAME record link to A record?

### DISCUSSION THREAD

**Discussion by jnc27** [sys4/netbox-plugin-dns#215] [Q&A] [answered]
How can a CNAME record linked to a A record? So that I know that these records belong together?

**Comment by peteeckel** [comment]
You can't, because they don't - at least not in a 1:1 relationship.

A CNAME record is not a link belonging to any particular record, but it's a kind of symbolic link the **name** of a set of records. Take the following example:

```
name1    IN    CNAME    name2

name2    IN     AAAA    fe80:dead:beef::1
         IN     A       10.0.0.1
```

Here, `name2` points to the **name** `name1`, not to any of the records sharing that name (there could be more than just `AAAA` and `A`). There is no obvious link from the `CNAME` to any of the records that happen to have the name the `CNAME` points to, so there is no way of attaching that kind of information to a `CNAME` in a static form.

What could be done is create a (dynamic) view panel in the `CNAME` detail view that lists the record it **currently** points to (and possibly vice versa) so you can see the relationship as it currently is.

**Reply by jnc27** [reply]
Can it (detail view) be done with the current version - or is this something you need to implement?
Thank you!

**Comment by peteeckel** [comment]
It's something I need to implement - I'm just looking into how it can be done.

**Reply by peteeckel** [reply]
Unfortunately there was a small bug in the code that prevented `CNAME`s on top level and `CNAME`s to top level records to be displayed (#220). This will be fixed in the next release.

**Comment by peteeckel** [comment]
<img width="1336" alt="Screenshot 2024-03-22 at 11 56 38" src="https://github.com/peteeckel/netbox-plugin-dns/assets/6815386/8b89cb7b-9b65-4192-8bbd-a96aa7528c21">

That's one direction, looking at the other one now ...

**Comment by peteeckel** [comment, marked as answer]
And the other direction works as well:
<img width="1251" alt="Screenshot 2024-03-22 at 12 50 07" src="https://github.com/peteeckel/netbox-plugin-dns/assets/6815386/23f9e347-0d7d-43d4-b319-edacaf80fcf2">

Now some tests and the new feature is ready to go :-)

**Reply by jnc27** [reply]
nice work! thank you!

**Reply by peteeckel** [reply]
Thank you for the new feature idea - thinking about it now it's obvious that this is really useful.

**Reply by aaronhuff** [reply]
Thanks for working on this, I have a custom script that creates CNAMES for the Primary and OOB IP for any device and I will pivot it to using this plugin! I would love if you could find a wild to couple this like you did with the ip addresses.

**Reply by peteeckel** [reply]
Hi @aaronhuff, thanks - glad you find it useful (so do I for my own stuff :-))!

As for doing the same with Device OOB/Primary/Secondary addresses, I think you need to specify in more detail what you have in mind. As I see it at the moment a  part of it is already there and another is impossible because it's not well-defined.

1. Primary IPv4, Primary IPv6 and OOB are nothing but references to `ipam.IPAddress` objects, and these objects can already be linked to DNS records. The only thing that might be a bit less comfortable is that if you use the "Create IP Address" option in the Device edit dialog you can't directly link to the DNS A/AAAA records as it is not that easy to smuggle the (currently) required custom fields into that part of the form. But if you create the IP addresses up front and then **assign** them to the interface (and designate them as primary) you're fine.
2. As for the `CNAME` - from what name should it point to the address records? I'm completely in the dark what you mean - can you please provide an example?

**Reply by aaronhuff** [reply]
My example should help clarify. While I want a DNS record (and PTR) for all assigned ip address in my netbox environment (which your solution will help me do), the gap is that I also want device level records (CNAMES to the appropriate IP address DNS record). For now I will use a custom script on device event, that creates the CNAME record manually but this seems like a common type of use case for DNS  automation.

My typically workflow would be implement a change via devicehostname-oob and monitor it via devicehostname to rollback. Where'as the IP specific ones are derived from the assigned interface for traceroutes etc. My DNS is 100% automated/programmatic in that nature.

PS. I also added a custom relationship between the VRF and a DNS Zone, as the way of automatically knowing which DNS values to generate for my devices.

```
IP Addresses:
- 1.1.1.1/32 -> sample-device-loopback1 (A + PTR)
- 10.0.0.100/24 -> sample-device-mgmt0 (A + PTR)

Device: 
- sample-device
    PrimaryIP4: 1.1.1.1/32 -> CNAME of sample-device pointing to sample-device-loopback1
    OOB-IP: 10.0.0.100/24 -> CNAME of sample-device-oob pointing to sample-device-mgmt0
```

So specifically, the "managed record" functionality I'm potentially proposing is that for an IP addresses that are set as the Primary or OOB-IP of a device, that a CNAME record be created that points to it's hostname or hostname-oob (I suspect this suffix would need to be customizable). If the primary IP, OOB IP or hostname updates.. so would the managed CNAME record accordingly.

I'm not sure what the middle ground generic implementation of this sort of functionality would look like but I wanted to plant the bug!

Thanks again @peteeckel !

**Comment by peteeckel** [comment]
Ah, I see - that's a bit of a special use case that's not easy to generalise.

So we're clear about the IP addresses themselves, right? 

That leaves the `CNAME` task. I see a couple of problems:

1. The device name is not necessarily a valid name for a DNS record. That is actually the worst part: A record name not only needs consist of a single word, it also needs to validate as a DNS host name. That is probably the case in your environment (otherwise you wouldn't come up with this suggestion), but in general it doesn't hold true.
2. For IP addresses and address records, validation of the resulting DNS record is pretty simple - the IP address itself is by definiton a valid value for an `A`/`AAAA` record, and the name stands for itself - it has no other meaning - and can be validated and rejected easily if it isn't a legal name. For Device names that's completely different - and in addition you get the problem of determining in which zone the cname should be created, which again is probably the same zone as the IPv4 and IPv6 address record. But that's also not something one can use for a general solution. 
3. Then, you mentioned it, there is the problem of creating suffixes that can be used for creating the `CNAME`s. Again something that needs to be configurable and bears huge potential for conflicts with existing names - which is especially horrible with `CNAME`, as any clash with another record leads to unloadable zones.
4. And then there is the fact that your case, while I assume there are a handful of users with that kind of requirement, is a bit exotic ... there are some more pressing issues, and since a general solution is probably not possible I would rather not open that box.

**But:** You have a special problem to solve, and you have NetBox. Therefore there is a solution, and in this case (like so often) it's a custom script. It is, if you know the algorithms you're using well, easy to create a script that takes a device object, extracts the primary and OOB addresses from it, finds the corresponding DNS records (if you don't know how I can give you a simple code snippet), create  valid DNS names for your `CNAME`s, and then create them via NetBox DNS (again, if you need a code example just ask, it's easy to do). Then you create a couple of event rules that makes the script trigger (depends on your workflow, but device creation and modification are good starting points), and that should already do. 

This will not be a general solution (as I said I doubt that it can be generalised at all), but it will fit **your** use case and take the onus of creating the `CNAME`s off you.

**Reply by aaronhuff** [reply]
Yeah, I think we are on the same page :). Thanks for the info. Like you said, Netbox has all of the mechanisms in place for me to fully automate this with event rules and custom scripts etc.
